### PR TITLE
Add savestate checkpoints to verify that MEASURE and WRITE match

### PR DESCRIPTION
**Pull request by hrydgard**
Verifies that MEASURE and WRITE match in every detail. Did this because I got save state crashes in Ratchet & Clank, writing beyond the end of the buffer.

And indeed, I have a situation now where saveDataListCount in SaveDataParam::DoState differs (0 during measure, 5 during write), leading to different data lengths! 

I can get into this situation by loading a state just after booting up Ratchet, and then loading it again, resulting in a backup save. During that backup save, this happens, sometimes leading to a crash if there was no spare space at the end of the last memory page of the buffer.

Will debug and fix that before merging this, but at least this should result in future cases like this gracefully failing to save instead of crashing.

This seems to have something to do PSPSaveDialog::DoState which calls param.SetPspParam which seems to modify it!

### CODE REVIEW

File: Common/Serialize/Serializer.cpp [hunk 1]
```diff
@@ -33,8 +33,25 @@ enum class SerializeCompressType {
 
 static constexpr SerializeCompressType SAVE_TYPE = SerializeCompressType::ZSTD;
 
-PointerWrapSection PointerWrap::Section(const char *title, int ver) {
-	return Section(title, ver, ver);
+void PointerWrap::RewindForWrite(u8 *writePtr) {
+	_assert_(mode == MODE_MEASURE);
+	// Switch to writing mode and
```
Comment (unknownbrackets) [inline Common/Serialize/Serializer.cpp:38]: This comment might not have been

-[Unknown]

File: Common/Serialize/Serializer.h [hunk 2]
```diff
@@ -120,7 +139,13 @@ class PointerWrap
 	void DoMarker(const char *prevName, u32 arbitraryNumber = 0x42);
 
 private:
+	size_t Offset() const { return *ptr - ptrStart_; }
+
 	const char *firstBadSectionTitle_ = nullptr;
+	u8 *ptrStart_;
+	std::vector<SerializeCheckpoint> checkpoints_;
```
Comment (unknownbrackets) [inline Common/Serialize/Serializer.h:149]: Not sure if it matters much, but might be nice to `reserve()` this to some value near its typical size?  Though, this probably is quick compared to all the memory copying save state writing does, anyway.

Would be nice to rebase out the accidental pspautotests revert.

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]
Ah, cool.  I guess this means it's before savedata dialog shutdown, or something, but the memory was changed.

Accidentally reverted pspautotests in the first commit, though.

-[Unknown]

File: Common/Serialize/Serializer.h (hunk 2 above)
Comment (hrydgard) [reply Common/Serialize/Serializer.h:149]: True, it's worth doing a reserve here, I'll check about how many sections we have. And I'll obviously revert the revert.

**Review by unknownbrackets — APPROVED** [review]

**Pull request merged**